### PR TITLE
Typescript code quality tools

### PR DESCRIPTION
**Pull request by chrisblossom**
This PR tackles item number 3 in [#197](https://github.com/davidtheclark/cosmiconfig/pull/197#issuecomment-504700070).

Again, it will probably easiest to look at each commit separately to get an idea of what was changed.

If you want me to merge the commits before merging let me know.

### CODE REVIEW

File: src/createExplorer.ts [hunk 1]
```diff
@@ -255,10 +255,10 @@ class Explorer {
     content: string | null,
   ): Promise<CosmiconfigResult> {
     return Promise.resolve()
-      .then(() => {
+      .then(():Promise<LoadedFileContent> | LoadedFileContent  => {
```
Comment (davidtheclark) [inline src/createExplorer.ts]: Should there be a space after this colon?

File: src/createExplorer.ts [hunk 2]
```diff
@@ -21,6 +21,15 @@ const MODE_SYNC = 'sync';
...
 
+function nextDirUp(dir: string): string {
+  return path.dirname(dir);
+}
+
+function getExtensionDescription(filepath: string): string {
+  const ext = path.extname(filepath);
+  return ext ? `extension "${ext}"` : 'files without extensions';
+}
+
```
Comment (davidtheclark) [inline src/createExplorer.ts]: I'd prefer not to have an ESLint rule that forces this unnecessary change: I don't think it prevents any errors, and I do think it can sometimes improve readability to take advantage of function hoisting.

File: src/getPropertyByPath.ts [hunk 3]
```diff
@@ -5,7 +5,7 @@
 // understood in array paths.
 function getPropertyByPath(
   source: { [key: string]: unknown },
-  path: string | Array<string>,
+  path: string | string[],
```
Comment (davidtheclark) [inline src/getPropertyByPath.ts]: I prefer the generic syntax, `Array<string>`, because it stands out more clearly.

File: src/createExplorer.ts (hunk 1 above)
Comment (chrisblossom) [reply src/createExplorer.ts]: I ran prettier after all commits to minimize changes per commit. This comment is marked outdated so I’m going to assume prettier did it’s job. (I’m on mobile right now, I’ll check for sure when I’m on a computer.

File: src/getPropertyByPath.ts (hunk 3 above)
Comment (chrisblossom) [reply src/getPropertyByPath.ts]: This comes from their recommended preset. Should I override?

File: src/createExplorer.ts (hunk 2 above)
Comment (chrisblossom) [reply src/createExplorer.ts]: I think you are referring to not using a function before it is defined? Personally I disagree, but don’t have strong feelings / don’t think it matters for small projects.

This comes from their recommended preset. Should I override?

Comment (olsonpm) [reply src/createExplorer.ts]: i don't care about stylistic decisions such as this when it's a shared project, but personally i use function hoisting a lot to make my code more readable.  In my opinion it allows for the higher level code to be read first without needing to bother with the implementation of it.

File: src/getPropertyByPath.ts (hunk 3 above)
Comment (chrisblossom) [reply src/getPropertyByPath.ts]: Fixed via [`'@typescript-eslint/array-type': ['error', 'generic']`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/array-type.md) - https://github.com/davidtheclark/cosmiconfig/pull/202/commits/a9e53cce3df0dff34e43eea2b18e57be302c838a

File: src/createExplorer.ts (hunk 2 above)
Comment (chrisblossom) [reply src/createExplorer.ts]: I was able to get back at a computer and revert this change. I also set [`'@typescript-eslint/no-use-before-define': 'off'`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-use-before-define.md).

File: src/createExplorer.ts (hunk 1 above)
Comment (chrisblossom) [reply src/createExplorer.ts]: Just checked, this was fixed in https://github.com/davidtheclark/cosmiconfig/pull/202/commits/19fc5ef3fd8d1bde07aee619ea00bbe2ea804429.
